homeassistant/components/nws/__init__.py

```diff
@@ -40,6 +43,59 @@ async def async_setup(hass: HomeAssistant, config: dict):
     return True
 
 
+class NwsDataUpdateCoordinator(DataUpdateCoordinator):
+    """
+    NWS data update coordinator.
+
+    Implements faster data update intervals for failed updates and exposes a last successful update time.
+    """
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        logger: logging.Logger,
+        *,
+        name: str,
+        update_interval: datetime.timedelta,
+        failed_update_interval: datetime.timedelta,
+        update_method: Optional[Callable[[], Awaitable]] = None,
+        request_refresh_debouncer: Optional[debounce.Debouncer] = None,
+    ):
+        """Initialize NWS coordinator."""
+        super().__init__(
+            hass,
+            logger,
+            name=name,
+            update_interval=update_interval,
+            update_method=update_method,
+            request_refresh_debouncer=request_refresh_debouncer,
+        )
+        self.failed_update_interval = failed_update_interval
+        self.last_update_success_time = None
+
+    @callback
+    def _schedule_refresh(self) -> None:
```

Comment:
Overwriting a protected method seems risky as it could break in the future.

Reply:
If we cannot rename, I could override `async_refresh`.  It has **a lot* more logic however and is probably more at risk of being changed in some part.

https://github.com/home-assistant/core/blob/dev/homeassistant/helpers/update_coordinator.py#L126

However, What would be ideal is a method that is called for successful and another method for unsuccessful updates.  Then, integrations could extend either of those.  I did not go this route based on the feedback from the prior PR however.

Reply:
@MartinHjelmare When you have a spare moment, I'm not sure how to advise here.

Reply:
Several integrations override _async_update_data, which I think I could also use. This seems cleaner, but I will wait on the word about the protected method issue before proceeding as it is the same.
https://github.com/home-assistant/core/blob/16670a38a432e556cdd406becef9ac6b084fbc86/homeassistant/components/airly/__init__.py#L118

Reply:
Yeah, we should probably make the methods public (or abstract) that we want to allow overriding. Although as a child, if you want to play with a parent's things, you're always required to follow that parent's rules.

Reply:
I'll take this to mean that I can override _async_update_data?

Reply:
It's ok yes.